src/m_class.c

```diff
@@ -529,6 +551,46 @@ void class_addcreator(t_newmethod newmethod, t_symbol *s,
     va_end(ap);
     class_addmethod(pd_objectmaker, (t_method)newmethod, s,
         vec[0], vec[1], vec[2], vec[3], vec[4], vec[5]);
+    if (class_loadsym)
+    {
+            /* if we're loading an extern it might have been invoked by a
+            longer file name; in this case, prepend the directory to the alias. */
+        const char *loadstring = class_loadsym->s_name, *end;
+        if ((end = strrchr(loadstring, '/')))
+        {
+            char buf[MAXPDSTRING];
+            size_t dirlen = end - loadstring + 1, namelen = strlen(s->s_name);
+            if ((dirlen + namelen) < MAXPDSTRING)
+            {
+                strncpy(buf, loadstring, dirlen);
+                strcpy(buf + dirlen, s->s_name);
+                /* post("add alias: %s", buf); */
+                class_addmethod(pd_objectmaker, (t_method)newmethod,
+                    gensym(buf),
+                    vec[0], vec[1], vec[2], vec[3], vec[4], vec[5]);
+            }
+        }
+    }
+    if (class_extern_sym && class_newest)
+    {
+            /* if the extern name doesn't end with the class name
+            we might be a multi-object-per-binary library,
+            so we prepend the library name to the alias to
+            avoid possible nameclashes with other libraries. */
+        const char *loadstring = class_extern_sym->s_name,
+                *classname = class_newest->c_name->s_name;
+        size_t l1 = strlen(classname), l2 = strlen(loadstring);
+        if (l2 < l1 || strcmp(classname, loadstring + (l2 - l1)))
+        {
+            char buf[MAXPDSTRING];
+            buf[MAXPDSTRING-1] = 0;
+            snprintf(buf, MAXPDSTRING-1, "%s/%s", loadstring, s->s_name);
+            /* post("add alias: %s", buf); */
+            class_addmethod(pd_objectmaker, (t_method)newmethod,
+                gensym(buf),
+                vec[0], vec[1], vec[2], vec[3], vec[4], vec[5]);
+        }
+    }
```

Comment:
shouldn't that duplicate code go into a separate helper function?

Reply:
thanks, done!